File: src/p_3_creational_patterns/p_11_objectpool/ConnectionPool.java
package p_5_creational_patterns.p_11_objectpool;

import java.util.ArrayList;
import java.util.List;

public class ConnectionPool {

    static List<Connection> connectionList = new ArrayList<>();

    public ConnectionPool(){
        init();
    }

    public void init(){
        for (int i = 0; i < 3; i++)
            connectionList.add(new Connection());
    }

    static Connection getConnection(){
        if (connectionList.size() == 0)
            throw new RuntimeException("there is no object");

        Connection instance = connectionList.get(0);
        ConnectionPool.connectionList.remove(0);
        return instance;
    }

    static void release(Connection conn)
    {
        if ( conn != null)
            connectionList.add(conn);
    }

}

File: Net/03_creational_patterns/07_objectpooling_pattern/ConnectionPool.cs
namespace _07_objectpooling_pattern
{
    public class ConnectionPool
    {
        static List<Connection> connectionList = new();

        public ConnectionPool()
        {
            Init();
        }

        public void Init()
        {
            for (int i = 0; i < 3; i++)
                connectionList.Add(new Connection());
        }

        public static Connection getConnection()
        {
            if (connectionList.Count == 0)
                throw new Exception("there is no object");

            Connection instance = connectionList[0];
            ConnectionPool.connectionList.RemoveAt(0);
            return instance;
        }

        public static void release(Connection conn)
        {
            if (conn != null)
                connectionList.Add(conn);
        }
    }
}

